jina/executors/decorators.py

```diff
@@ -255,11 +274,79 @@ def arg_wrapper(*args, **kwargs):
+                         batch_size: Union[int, Callable] = None,
+                         num_batch: Optional[int] = None,
+                         split_over_axis: int = 0,
+                         merge_over_axis: int = 0,
+                         slice_on: int = 1,
+                         num_data: int) -> Any:
+    """Split the input of a function into small batches and call :func:`func` on each batch
+    , collect the merged result and return. This is useful when the input is too big to fit into memory
+
+    :param func: function to decorate
+    :param batch_size: size of each batch
+    :param num_batch: number of batches to take, the rest will be ignored
+    :param split_over_axis: split over which axis into batches
+    :param merge_over_axis: merge over which axis into a single result
+    :param slice_on: the location of the data. When using inside a class,
+            ``slice_on`` should take ``self`` into consideration.
+    :param num_data: the number of data inside the arguments (starting from ``slice_on``)
+    :return: the merged result as if run :func:`func` once on the input.
+
+    Example:
+        .. highlight:: python
+        .. code-block:: python
+
+            class MultiModalExecutor:
+
+                @batching_multi_input(batch_size = 64, num_data=2)
+                def encode(self, *batches, **kwargs):
+                    batch_modality0 = batches[0]
+                    embed0 = _encode_modality(batch_modality0)
+                    batch_modality1 = batches[1]
+                    embed1 = _encode_modality(batch_modality0)
+    """
+
+    def _batching(func):
+        @wraps(func)
+        def arg_wrapper(*args, **kwargs):
+            # priority: decorator > class_attribute
+            # by default data is in args[1:] (self needs to be taken into account)
+            data_args = [args[slice_on + i] for i in range(0, num_data)]
+
+            data = list(zip(*data_args))
+            args = list(args)
+
+            b_size = batch_size or getattr(args[0], 'batch_size', None)
+            # no batching if b_size is None
+            if b_size is None:
+                return func(*args, **kwargs)
+
+            default_logger.info(
+                f'batching enabled for {func.__qualname__} batch_size={b_size} '
+                f'num_batch={num_batch} axis={split_over_axis}')
+
+            # assume all datas have the same length
+            total_size = _get_total_size(data, b_size, num_batch, split_over_axis)
+            final_result = []
+
+            for multiple_batch in batch_iterator(data[:total_size], b_size, split_over_axis):
+                multiple_batch = list(zip(*multiple_batch))
+
+                for idx, data_batch in enumerate(multiple_batch):
+                    args[slice_on + idx] = data_batch
+
+                r = func(*args, **kwargs)
+
+                if r is not None:
+                    final_result.append(r)
```

Comment:
append or extend? Not sure, was just wondering

Reply:
I would break these 2 methods into smaller parts to share logic, but right now `single batching` has grown really into a big giant